// @effect/io/mjs/RequestBlock.mjs
import * as _RequestBlock from "@effect/io/internal_effect_untraced/blockedRequests";
import * as core from "@effect/io/internal_effect_untraced/core";
import * as _dataSource from "@effect/io/internal_effect_untraced/dataSource";
/**
 * @since 1.0.0
 * @category constructors
 */
export const single = _RequestBlock.single;
/**
 * @since 1.0.0
 * @category constructors
 */
export const empty = _RequestBlock.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
export const mapRequestResolvers = _RequestBlock.mapRequestResolvers;
/**
 * @since 1.0.0
 * @category constructors
 */
export const parallel = _RequestBlock.par;
/**
 * @since 1.0.0
 * @category constructors
 */
export const reduce = _RequestBlock.reduce;
/**
 * @since 1.0.0
 * @category constructors
 */
export const sequential = _RequestBlock.seq;
/**
 * Provides each data source with part of its required environment.
 *
 * @since 1.0.0
 * @category utils
 */
export const contramapContext = (self, f) => reduce(self, ContramapContextReducer(f));
const ContramapContextReducer = f => ({
  emptyCase: () => empty,
  parCase: (left, right) => parallel(left, right),
  seqCase: (left, right) => sequential(left, right),
  singleCase: (dataSource, blockedRequest) => single(_dataSource.contramapContext(dataSource, f), blockedRequest)
});
/**
 * Provides each data source with a fiber ref value.
 *
 * @since 1.0.0
 * @category utils
 */
export const locally = core.requestBlockLocally;
